include/vcmi/spells/Caster.h

```diff
@@ -68,9 +69,9 @@ class DLL_LINKAGE Caster
 	virtual void spendMana(ServerCallback * server, const int32_t spellCost) const = 0;
 
 	virtual int32_t manaLimit() const = 0;
-	
-	///used to identify actual hero caster
-	virtual const CGHeroInstance * getHeroCaster() const = 0;
+
+	/// This caster ID should be used only in adventure spells (-1 if no adventure casting possible)
+	virtual int32_t getAdventureCasterID() const = 0;
```

Comment:
- Why const CGHeroInstance * got replaced with int32_t?
- What this 'int32_t' means? ObjectInstanceID? HeroTypeID? Something else?

IMO preference for object identification should be:
- Pointer to corresponding object
- enumeration with index of corresponding object
- integer type
Unless there is strong reason not to use pointer/enum

Reply:
@IvanSavenko this is some id, for hero - it is HeroInstanceID, but this a caster interface, and I do not think than it is good idea to use CGHeroInstance * in spells::Caster (which should not know about heroes or units at all).

Reply:
I get the idea, but for developer who uses caster interface this won't be as clear....

It also drops any type-safety - if somebody accidentally uses wrong type here compliler won't generate an error.

Perhaps add 
using AdventureMapCaster = CGHeroInstance
in caster interface?

Or make CGHeroInstance inherit from some minimalistic interface, e.g. `AdventureMapCaster`?

Reply:
Then we need to make BattleCasterID return CreatureID too, and caster will not free of assumption is there units of creatures. I think we should return an abstract ID in both cases and do not assume what ID is.

Reply:
creating tiny interfaces (or perhaps even empty "tags") and making creature / hero inherit from them might be better approach then:
- caster class still has no knowledge of concrete types
- compiler can perform type checks since you can't dynamic_cast unrelated types

Reply:
@IvanSavenko my idea is we should not assume than caster is some concrete type, even tag. I want to do adventure spell refactoring, but not in this PR.

Reply:
But you already making such assumption by returning ID to *some* object. Meaning, you imply that caster is a type that has some sort of index.

If we don't want to make such assumptions then we should not provide such method to begin with.

Reply:
We can return ObjectInstanceID from getAdventureCasterID, but, I think, not more. And we should make sure than ObjectInstanceID has NONE.

Reply:
Sounds good. And if ObjectInstanceID::NONE does not exists - then feel free to add as '-1'

Reply:
done